File: backend/src/productflow_backend/application/media_library/queries.py
"""全局图库列表与 bootstrap。

按 MediaLibraryAsset 身份分页；文件夹/标签/未整理是组织投影，provenance 不进入列表载荷。
"""

from __future__ import annotations

import base64
import binascii
import hashlib
import json
from dataclasses import dataclass
from datetime import UTC, datetime
from typing import Any

from sqlalchemy import and_, func, or_, select
from sqlalchemy.orm import Session, defer, selectinload

from productflow_backend.application.media_library.contracts import MediaLibrarySourceType
from productflow_backend.domain.errors import BusinessValidationError, NotFoundError
from productflow_backend.infrastructure.db.models import (
    MediaLibraryAsset,
    MediaLibraryAssetTag,
    MediaLibraryFolder,
    MediaLibraryTag,
)


def _asset_query():
    return select(MediaLibraryAsset).options(
        defer(MediaLibraryAsset.provenance_json),
        selectinload(MediaLibraryAsset.media_object),
        selectinload(MediaLibraryAsset.source_image_session_asset),
        selectinload(MediaLibraryAsset.source_product_asset),
        selectinload(MediaLibraryAsset.folder),
        selectinload(MediaLibraryAsset.tag_assignments).selectinload(MediaLibraryAssetTag.tag),
    )


@dataclass(frozen=True, slots=True)
class MediaLibraryAssetPage:
    items: tuple[MediaLibraryAsset, ...]
    next_cursor: str | None


@dataclass(frozen=True, slots=True)
class MediaLibraryBootstrap:
    total_count: int
    active_count: int
    archived_count: int
    unorganized_count: int
    folders: tuple[tuple[str, str, int], ...]
    tags: tuple[tuple[str, str, int], ...]


def get_media_library_asset(session: Session, *, asset_id: str) -> MediaLibraryAsset:
    """按全局素材 id 读取一条资产及其 MediaObject。"""

    asset = session.scalar(_asset_query().where(MediaLibraryAsset.id == asset_id))
    if asset is None:
        raise NotFoundError("素材库资产不存在")
    return asset


_CURSOR_VERSION = 2


def _normalize_search(search: str | None) -> str:
    return " ".join((search or "").split())


def _cursor_filter_signature(
    *,
    search: str,
    source_type: MediaLibrarySourceType | None,
    include_archived: bool,
    folder_id: str | None,
    tag: str | None,
) -> str:
    payload = json.dumps(
        {
            "folder_id": folder_id or "",
            "include_archived": include_archived,
            "search": search,
            "source_type": source_type or "",
            "tag": tag or "",
        },
        sort_keys=True,
        separators=(",", ":"),
    )
    return hashlib.sha256(payload.encode("utf-8")).hexdigest()


def _coerce_utc(value: datetime) -> datetime:
    if value.tzinfo is None:
        return value.replace(tzinfo=UTC)
    return value.astimezone(UTC)


def _encode_cursor(*, created_at: datetime, asset_id: str, signature: str, as_of: datetime) -> str:
    payload = {
        "v": _CURSOR_VERSION,
        "sort": "created_desc",
        "filter": signature,
        "key": _coerce_utc(created_at).isoformat(),
        "id": asset_id,
        "as_of": _coerce_utc(as_of).isoformat(),
    }
    encoded = json.dumps(payload, sort_keys=True, separators=(",", ":")).encode()
    return base64.urlsafe_b64encode(encoded).decode().rstrip("=")


def _decode_cursor(cursor: str, *, signature: str) -> tuple[datetime, str, datetime]:
    try:
        padded = cursor + "=" * (-len(cursor) % 4)
        payload: Any = json.loads(base64.urlsafe_b64decode(padded.encode("ascii")))
        if (
            not isinstance(payload, dict)
            or payload.get("v") != _CURSOR_VERSION
            or payload.get("sort") != "created_desc"
            or payload.get("filter") != signature
            or not all(isinstance(payload.get(key), str) and payload[key] for key in ("key", "id", "as_of"))
        ):
            raise ValueError
        created_at = datetime.fromisoformat(payload["key"])
        as_of = datetime.fromisoformat(payload["as_of"])
        if created_at.tzinfo is None:
            created_at = created_at.replace(tzinfo=UTC)
        if as_of.tzinfo is None:
            as_of = as_of.replace(tzinfo=UTC)
        return created_at, payload["id"], as_of
    except (ValueError, TypeError, UnicodeDecodeError, json.JSONDecodeError, binascii.Error) as exc:
        raise BusinessValidationError("素材库分页游标无效或与当前筛选条件不匹配") from exc


def _filtered_asset_statement(
    *,
    search: str,
    source_type: MediaLibrarySourceType | None,
    include_archived: bool,
    folder_id: str | None,
    tag: str | None,
):
    statement = select(MediaLibraryAsset)
    if not include_archived:
        statement = statement.where(MediaLibraryAsset.is_archived.is_(False))
    if search:
        pattern = f"%{search}%"
        statement = statement.where(
            or_(MediaLibraryAsset.display_name.ilike(pattern), MediaLibraryAsset.original_filename.ilike(pattern))
        )
    if source_type:
        statement = statement.where(MediaLibraryAsset.source_type == source_type)
    if folder_id:
        statement = statement.where(MediaLibraryAsset.folder_id == folder_id)
    if tag:
        statement = statement.join(MediaLibraryAssetTag, MediaLibraryAssetTag.asset_id == MediaLibraryAsset.id).join(
            MediaLibraryTag, MediaLibraryTag.id == MediaLibraryAssetTag.tag_id
        ).where(MediaLibraryTag.normalized_name == tag)
    return statement


def get_media_library_bootstrap(session: Session) -> MediaLibraryBootstrap:
    """活跃/归档/未整理与一层文件夹计数，都是查询投影。"""

    total_count = session.scalar(select(func.count(MediaLibraryAsset.id))) or 0
    active_count = (
        session.scalar(
            select(func.count(MediaLibraryAsset.id)).where(MediaLibraryAsset.is_archived.is_(False))
        )
        or 0
    )
    archived_count = total_count - active_count
    unorganized_count = session.scalar(
        select(func.count(MediaLibraryAsset.id)).where(
            MediaLibraryAsset.is_archived.is_(False), MediaLibraryAsset.folder_id.is_(None)
        )
    ) or 0
    folder_rows = session.execute(
        select(MediaLibraryFolder.id, MediaLibraryFolder.name, func.count(MediaLibraryAsset.id))
        .outerjoin(
            MediaLibraryAsset,
            (MediaLibraryAsset.folder_id == MediaLibraryFolder.id)
            & MediaLibraryAsset.is_archived.is_(False),
        )
        .group_by(MediaLibraryFolder.id)
        .order_by(MediaLibraryFolder.name, MediaLibraryFolder.id)
    ).all()
    tag_rows = session.execute(
        select(MediaLibraryTag.id, MediaLibraryTag.name, func.count(MediaLibraryAsset.id))
        .outerjoin(MediaLibraryAssetTag, MediaLibraryAssetTag.tag_id == MediaLibraryTag.id)
        .outerjoin(
            MediaLibraryAsset,
            (MediaLibraryAsset.id == MediaLibraryAssetTag.asset_id)
            & MediaLibraryAsset.is_archived.is_(False),
        )
        .group_by(MediaLibraryTag.id)
        .order_by(MediaLibraryTag.name, MediaLibraryTag.id)
    ).all()
    return MediaLibraryBootstrap(
        total_count=total_count,
        active_count=active_count,
        archived_count=archived_count,
        unorganized_count=unorganized_count,
        folders=tuple(folder_rows),
        tags=tuple(tag_rows),
    )


def list_media_library_assets(
    session: Session,
    *,
    limit: int = 20,
    cursor: str | None = None,
    include_archived: bool = False,
    search: str | None = None,
    source_type: MediaLibrarySourceType | None = None,
    folder_id: str | None = None,
    tag: str | None = None,
) -> MediaLibraryAssetPage:
    """按全局素材身份分页；文件夹/标签只是筛选投影。"""

    bounded_limit = min(max(limit, 1), 100)
    normalized_search = _normalize_search(search)
    normalized_tag = tag.casefold().strip() if tag else None
    signature = _cursor_filter_signature(
        search=normalized_search,
        source_type=source_type,
        include_archived=include_archived,
        folder_id=folder_id.strip() if folder_id else None,
        tag=normalized_tag,
    )
    as_of = datetime.now(UTC)
    if cursor:
        cursor_created_at, cursor_id, as_of = _decode_cursor(cursor, signature=signature)
    else:
        cursor_created_at = cursor_id = None
    statement = _filtered_asset_statement(
        search=normalized_search,
        source_type=source_type,
        include_archived=include_archived,
        folder_id=folder_id.strip() if folder_id else None,
        tag=normalized_tag,
    ).options(
        defer(MediaLibraryAsset.provenance_json),
        selectinload(MediaLibraryAsset.media_object),
        selectinload(MediaLibraryAsset.source_image_session_asset),
        selectinload(MediaLibraryAsset.source_product_asset),
        selectinload(MediaLibraryAsset.folder),
        selectinload(MediaLibraryAsset.tag_assignments).selectinload(MediaLibraryAssetTag.tag),
    ).order_by(MediaLibraryAsset.created_at.desc(), MediaLibraryAsset.id.desc()).limit(bounded_limit + 1)
    if cursor_created_at is not None:
        statement = statement.where(
            or_(
                MediaLibraryAsset.created_at < cursor_created_at,
                and_(MediaLibraryAsset.created_at == cursor_created_at, MediaLibraryAsset.id < cursor_id),
            )
        )
    rows = list(session.scalars(statement).unique().all())
    has_more = len(rows) > bounded_limit
    items = tuple(rows[:bounded_limit])
    next_cursor = (
        _encode_cursor(created_at=items[-1].created_at, asset_id=items[-1].id, signature=signature, as_of=as_of)
        if has_more and items
        else None
    )
    return MediaLibraryAssetPage(items=items, next_cursor=next_cursor)

File: go/internal/library/queries.go
package library

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"strings"
	"time"

	"github.com/yuqie6/productflow/internal/platform/apperr"
	"github.com/yuqie6/productflow/internal/platform/db/schema"
	"github.com/yuqie6/productflow/internal/platform/tx"
	"gorm.io/gorm"
)

// Get 按 id 读取全局素材身份（内部 Asset，不是 AssetResponse）。
// 调用时机：HTTP GET/download/archive/restore，以及收录前核验。
// 找不到返回 NotFound「素材库资产不存在」。不读磁盘 bytes；不要当工作流子图库关联用。
func (s Service) Get(ctx context.Context, assetID string) (Asset, error) {
	var asset Asset
	err := tx.WithGorm(ctx, s.DB, func(pgxTx *gorm.DB) error {
		var err error
		asset, err = s.loadAsset(ctx, pgxTx, assetID)
		return err
	})
	return asset, err
}

// Bootstrap 读图库首页计数、一层文件夹与标签，无写入。
// 调用时机：HTTP GET /bootstrap。文件夹 Count 只含未归档素材。
// 读库失败或 ctx 取消时返回 error。
func (s Service) Bootstrap(ctx context.Context) (Bootstrap, error) {
	var out Bootstrap
	err := tx.WithGorm(ctx, s.DB, func(pgxTx *gorm.DB) error {
		var total, active, unorganized int64
		if err := pgxTx.WithContext(ctx).Model(&schema.MediaLibraryAssets{}).Count(&total).Error; err != nil {
			return err
		}
		if err := pgxTx.WithContext(ctx).Model(&schema.MediaLibraryAssets{}).Where("is_archived = ?", false).Count(&active).Error; err != nil {
			return err
		}
		out.TotalCount = int(total)
		out.ActiveCount = int(active)
		out.ArchivedCount = out.TotalCount - out.ActiveCount
		if err := pgxTx.WithContext(ctx).Model(&schema.MediaLibraryAssets{}).
			Where("is_archived = ? AND folder_id IS NULL", false).Count(&unorganized).Error; err != nil {
			return err
		}
		out.UnorganizedCount = int(unorganized)
		var folders []struct {
			ID    string `gorm:"column:id"`
			Name  string `gorm:"column:name"`
			Count int64  `gorm:"column:count"`
		}
		if err := pgxTx.WithContext(ctx).Table("media_library_folders AS f").
			Select("f.id, f.name, COUNT(a.id) AS count").
			Joins("LEFT JOIN media_library_assets a ON a.folder_id = f.id AND a.is_archived = FALSE").
			Group("f.id").
			Order("f.name, f.id").
			Scan(&folders).Error; err != nil {
			return err
		}
		out.Folders = make([]Folder, 0, len(folders))
		for _, folder := range folders {
			out.Folders = append(out.Folders, Folder{ID: folder.ID, Name: folder.Name, Count: int(folder.Count)})
		}
		var tags []struct {
			ID    string `gorm:"column:id"`
			Name  string `gorm:"column:name"`
			Count int64  `gorm:"column:count"`
		}
		if err := pgxTx.WithContext(ctx).Table("media_library_tags AS t").
			Select("t.id, t.name, COUNT(a.id) AS count").
			Joins("LEFT JOIN media_library_asset_tags at ON at.tag_id = t.id").
			Joins("LEFT JOIN media_library_assets a ON a.id = at.asset_id AND a.is_archived = FALSE").
			Group("t.id").
			Order("t.name, t.id").
			Scan(&tags).Error; err != nil {
			return err
		}
		out.Tags = make([]Tag, 0, len(tags))
		for _, tag := range tags {
			out.Tags = append(out.Tags, Tag{ID: tag.ID, Name: tag.Name, Count: int(tag.Count)})
		}
		return nil
	})
	return out, err
}

// List 按 ListFilter 分页列出全局素材 HTTP 投影。
// 调用时机：HTTP GET /api/media-library。Limit 默认 20、上限 100。
// 游标与当前筛选签名不一致返回 Validation，禁止跨筛选续页。
func (s Service) List(ctx context.Context, in ListFilter) (ListResponse, error) {
	if in.Limit < 1 {
		in.Limit = 20
	}
	if in.Limit > 100 {
		in.Limit = 100
	}
	search := normalizeSearch(in.Search)
	sourceType := strings.TrimSpace(in.SourceType)
	folderID := strings.TrimSpace(in.FolderID)
	tag := ""
	if strings.TrimSpace(in.Tag) != "" {
		tag = strings.TrimSpace(in.Tag)
	}
	signature := filterSignature(search, sourceType, in.IncludeArchived, folderID, tag)
	asOf := s.now()
	var cursorCreated *time.Time
	var cursorID string
	if strings.TrimSpace(in.Cursor) != "" {
		created, id, decodedAsOf, err := decodeCursor(in.Cursor, signature)
		if err != nil {
			return ListResponse{}, err
		}
		cursorCreated = &created
		cursorID = id
		asOf = decodedAsOf
	}
	var page ListResponse
	err := tx.WithGorm(ctx, s.DB, func(pgxTx *gorm.DB) error {
		q := libraryAssetQuery(pgxTx.WithContext(ctx))
		if !in.IncludeArchived {
			q = q.Where("a.is_archived = ?", false)
		}
		if search != "" {
			pattern := "%" + search + "%"
			q = q.Where("(a.display_name ILIKE ? OR a.original_filename ILIKE ?)", pattern, pattern)
		}
		if sourceType != "" {
			q = q.Where("a.source_type = ?", sourceType)
		}
		if folderID != "" {
			q = q.Where("a.folder_id = ?", folderID)
		}
		if tag != "" {
			q = q.Joins("JOIN media_library_asset_tags at ON at.asset_id = a.id").
				Joins("JOIN media_library_tags tg ON tg.id = at.tag_id").
				Where("tg.normalized_name = ?", tag)
		}
		if cursorCreated != nil {
			q = q.Where("(a.created_at < ? OR (a.created_at = ? AND a.id < ?))", *cursorCreated, *cursorCreated, cursorID)
		}
		q = q.Order("a.created_at DESC, a.id DESC").Limit(in.Limit + 1)
		items, err := loadAssets(ctx, pgxTx, q)
		if err != nil {
			return err
		}
		hasMore := len(items) > in.Limit
		if hasMore {
			items = items[:in.Limit]
		}
		page.Items, err = serializeAssets(items)
		if err != nil {
			return err
		}
		if hasMore && len(items) > 0 {
			next, err := encodeCursor(items[len(items)-1].CreatedAt, items[len(items)-1].ID, signature, asOf)
			if err != nil {
				return err
			}
			page.NextCursor = &next
		}
		return nil
	})
	return page, err
}

func encodeCursor(createdAt time.Time, assetID, signature string, asOf time.Time) (string, error) {
	raw, err := canonicalJSON(map[string]any{
		"v":      2,
		"sort":   "created_desc",
		"filter": signature,
		"key":    pythonISOFormat(createdAt),
		"id":     assetID,
		"as_of":  pythonISOFormat(asOf),
	})
	if err != nil {
		return "", err
	}
	return strings.TrimRight(base64.URLEncoding.EncodeToString(raw), "="), nil
}

// decodeCursor 解开 v2 游标。sort/filter 与当前列表签名不一致时一律 400，避免跨筛选续页。
func decodeCursor(cursor, signature string) (time.Time, string, time.Time, error) {
	invalid := apperr.Validation("素材库分页游标无效或与当前筛选条件不匹配")
	padded := cursor + strings.Repeat("=", (4-len(cursor)%4)%4)
	raw, err := base64.URLEncoding.DecodeString(padded)
	if err != nil {
		return time.Time{}, "", time.Time{}, invalid
	}
	var payload map[string]any
	if err := json.Unmarshal(raw, &payload); err != nil {
		return time.Time{}, "", time.Time{}, invalid
	}
	if asIntOr(payload["v"], 0) != 2 || payload["sort"] != "created_desc" || payload["filter"] != signature {
		return time.Time{}, "", time.Time{}, invalid
	}
	key, _ := payload["key"].(string)
	id, _ := payload["id"].(string)
	asOfRaw, _ := payload["as_of"].(string)
	if key == "" || id == "" || asOfRaw == "" {
		return time.Time{}, "", time.Time{}, invalid
	}
	created, err := parseDateTime(key)
	if err != nil {
		return time.Time{}, "", time.Time{}, invalid
	}
	asOf, err := parseDateTime(asOfRaw)
	if err != nil {
		return time.Time{}, "", time.Time{}, invalid
	}
	return created, id, asOf, nil
}

func asIntOr(v any, fallback int) int {
	n, err := asInt(v)
	if err != nil {
		return fallback
	}
	return n
}
